packages/hydrogen/src/analytics-manager/AnalyticsProvider.tsx

```diff
@@ -0,0 +1,524 @@
+import {
+  type ReactNode,
+  useEffect,
+  useState,
+  useMemo,
+  createContext,
+  useContext,
+  useRef,
+} from 'react';
+import {type CartReturn} from '../cart/queries/cart-types';
+import {
+  AnalyticsPageView,
+  AnalyticsProductView,
+  AnalyticsCollectionView,
+  AnalyticsCartView,
+  AnalyticsSearchView,
+  AnalyticsCustomView,
+  type PageViewPayload,
+  type ProductViewPayload,
+  type CollectionViewPayload,
+  type CartViewPayload,
+  type CartUpdatePayload,
+  type CustomEventPayload,
+  type OtherData,
+  type EventPayloads,
+  type CartLineUpdatePayload,
+  type SearchViewPayload,
+} from './AnalyticsView';
+import type {
+  CurrencyCode,
+  LanguageCode,
+  Shop,
+  Localization,
+} from '@shopify/hydrogen-react/storefront-api-types';
+import {AnalyticsEvent} from './events';
+import {ShopifyAnalytics} from './ShopifyAnalytics';
+import {CartAnalytics} from './CartAnalytics';
+import type {CustomerPrivacyApiProps} from '../customer-privacy/ShopifyCustomerPrivacy';
+import type {Storefront} from '../storefront';
+
+export type ShopAnalytic = {
+  /** The shop ID. */
+  shopId: string;
+  /** The language code that is being displayed to user. */
+  acceptedLanguage: LanguageCode;
+  /** The currency code that is being displayed to user. */
+  currency: CurrencyCode;
+  /** The Hydrogen subchannel ID generated by Oxygen in the environment variable. */
+  hydrogenSubchannelId: string | '0';
+};
+
+export type AnalyticsProviderProps = {
+  /** React children to render. */
+  children?: ReactNode;
+  /** The cart or cart promise to track for cart analytics. When there is a difference between the state of the cart, `AnalyticsProvider` will trigger a `cart_updated` event. It will also produce `product_added_to_cart` and `product_removed_from_cart` based on cart line quantity and cart line id changes. */
+  cart: Promise<CartReturn | null> | CartReturn | null;
+  /** An optional function to set wether the user can be tracked. Defaults to Customer Privacy API's `window.Shopify.customerPrivacy.userCanBeTracked()`. */
+  canTrack?: () => boolean;
+  /** An optional custom payload to pass to all events. e.g language/locale/currency. */
+  customData?: Record<string, unknown>;
```

Comment:
Would it be useful to allow a function here instead / in addition to the object?
Mostly to return different type of customData depending on the event. Or maybe the function is passed the whole event data and they can modify it?

Just a random thought, not really asking for this feature since I'm not sure how common it is.

Reply:
mmm you mean the function would be behaving like a middleware (for analytics)? I feel that would cause confusion of where the data is being modified/enhanced?

Reply:
yea it shouldn't mutate across events - if they need something specific for that event, they can manipulate at where they subscribe to the event